prov/rxm/src/rxm_init.c

```diff
@@ -284,6 +284,14 @@ RXM_INI
 			"Messages of size greater than this (default: 256 Kb) "
 			"would be transmitted via rendezvous protocol.");
 
+	fi_param_define(&rxm_prov, "sar_min_seg_num", FI_PARAM_SIZE_T,
+			"Defines the minimum number of segments that are used "
+			"in SAR protocol (default: 4).");
```

Comment:
I don't understand the need for a minimum segment number.   I would have expected SAR to be controlled through a single value -- either "maximum SAR size" or "minimum rendezvous size".  Anything over that would use rendezvous.  Anything less would use SAR, with special handling when the number of segments is simply 1.

Reply:
"maximum SAR size" is FI_OFI_RXM_SAR_LIMIT.
But we need to ensure that data is segmented by SAR protocol in a most optimal way
For example, it's more optimal to split 16-32 Kb into small segments - 4k-8k, but for large messages, we use maximum buffer size, i.e. eager size.

Reply:
Why would we want to split a message into smaller segments than the eager size?  That's counter-intuitive to me.  More importantly, why expose this to a user?

Reply:
It's may be more than platform and also related to the total message size, plus fabric characteristics.  My guess is that the benefit comes from overlapping the data copy with the posting of the buffers, so that a fairly constant stream of data is flowing.  But this is a fairly complex algorithm that I think we should keep internal, rather than trying to expose through environment variables.

My vote is to only expose a single variable to control when to switch from an 'eager' mode (including SAR) to rendezvous.  For eager/SAR, we only implement SAR.  As a start, I would implement SAR as:

1st packet: 1k (or 2k)
2nd packet: 2k (or 4k)
3rd packet: 4k (or 8K)
etc. up to the max eager msg size.

This algorithm worked well with rsockets.  If we end up with multiple algorithms, for example add an option to send all packets as the same size, then we could look at exposing additional variables to select the algorithm and fine tune it.

Reply:
Yes, the idea of SAR - overlapping local memory copy and communication part of protocol.

> As a start, I would implement SAR as:

I like your idea, it should help. just to confirm, Did you mean to split a message into segments by the following formula: `SUM(pow(2, N) * 1k; N) + Remaining`, where N - number of segment = `{0, 1, 2 ...}` and Remaining - a segment that has size = `Message Size - SUM(pow(2, N) * 1k; N)`
for example, 64514 = 1024 + 2048 + 4096 + 8192 + 16384 + 16384 + 2. 16384 - eager msg size

Reply:
Yes, something like this.  Whether you start at 1k or 2k is up to you.  The remaining part is a little hard to deal with.  Maybe if the remaining size is < 1k, toss it in with the first or second packet?  I'm not sure how to quickly calculate what the remaining size would be.

Reply:
> Maybe if the remaining size is < 1k, toss it in with the first or second packet? I'm not sure how to quickly calculate what the remaining size would be.

Do you mean to send the remaining part with the 1st or 2nd? I think we should have a if statement on a receiver to check whether this message contains two segment or not.

Reply:
I meant that, but only second thought, it's probably not worth the effort, especially now.  It only has the potential to help in very edge cases.